# IMPORTS

In [3]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
import matplotlib.pyplot as plt
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
from sklearn.model_selection import GridSearchCV, train_test_split
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_predict
from sklearn.model_selection import cross_val_score
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import confusion_matrix
from xgboost import XGBClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.cluster import KMeans
from sklearn.discriminant_analysis import StandardScaler
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
from sklearn.svm import SVC
from sklearn.feature_selection import VarianceThreshold
import matplotlib.pyplot as plt
import numpy as np
from sklearn.metrics import f1_score



## LEITURA DOS DADOS

In [4]:
df = pd.read_csv('/home/cid34senhas/Desktop/DAATP/Entrega/1_DF_CN-MCI_1_After_Pre_Processamento.csv') 
df_test = pd.read_csv('/home/cid34senhas/Desktop/DAATP/Entrega/1_DF_Test_1_After_Pre_Processamento.csv')






# Tratamento

### Desvio Padrão

In [5]:

def remove_low_variance_features(data, threshold):
    """
    Remove columns with a standard deviation below the given threshold.

    Parameters:
        data (pd.DataFrame): The input DataFrame.
        threshold (float): The minimum standard deviation a column must have to be retained.

    Returns:
        tuple: A tuple containing:
            - pd.DataFrame: A DataFrame with low-variance columns removed.
            - list: A list of columns that were removed.
    """
    # Calculate the standard deviation for each column
    std_dev = data.std()

    # Identify columns to keep (standard deviation above threshold)
    retained_columns = std_dev[std_dev > threshold].index

    # Identify columns to remove
    removed_columns = std_dev[std_dev <= threshold].index

    # Create a new DataFrame with only the retained columns
    data_reduced = data[retained_columns]

    return data_reduced, removed_columns

In [6]:
# *!  df.to_csv('DF_1_After_Pre_Processamento.csv', index=False)
# *! ## df_test.to_csv('DF_Test_1_After_Pre_Processamento.csv', index=False)


### Funcao Axuliar: Matrix de Confusão

In [7]:
def plot_confusion_matrix_with_labels(confusion_matrix, label_mapping):
    # Criar um mapeamento inverso
    reverse_label_mapping = {v: k for k, v in label_mapping.items()}
    
    # Aplicar o mapeamento inverso na matriz de confusão
    cm_with_labels = np.zeros_like(confusion_matrix, dtype=object)
    for i in range(confusion_matrix.shape[0]):
        for j in range(confusion_matrix.shape[1]):
            cm_with_labels[i, j] = f"{reverse_label_mapping[i]} (Pred: {reverse_label_mapping[j]})"

    # Criar um DataFrame para facilitar a visualização
    df_cm = pd.DataFrame(confusion_matrix, index=reverse_label_mapping.values(), columns=reverse_label_mapping.values())
    
    # Plotar a matriz de confusão
    plt.figure(figsize=(10, 7))
    sns.heatmap(df_cm, annot=True, fmt="d", cmap="Blues")
    plt.xlabel('Predicted')
    plt.ylabel('True')
    plt.title('Confusion Matrix with Labels')
    plt.show()

# MODELOS

Função que testa todos os nossos modelos que estamos a analisar, consuante um certo dataset

### Geração Dos Vários datasets para os vários modelos 

In [8]:
#data_ExtraTrees,_ = remove_highly_correlated_features(df,'Transition',0.93)
#data_Random_Forest,_ = remove_highly_correlated_features(df,'Transition',0.81)
#data_GradientBoost,_ = remove_highly_correlated_features(df,'Transition',0.83)
#data_XGBoost,_ = remove_highly_correlated_features(df,'Transition',0.82)
#
#data_ExtraTrees.to_csv('2_data_ExtraTrees.csv', index=False)
#data_Random_Forest.to_csv('2_data_Random_Forest.csv', index=False)
#data_GradientBoost.to_csv('2_data_GradientBoost.csv', index=False)
#data_XGBoost.to_csv('2_data_XGBoost.csv', index=False)
#

In [9]:
#X = df.drop('Transition', axis=1)
#y = df['Transition']
#
#_ = train_and_predict(X,y)

In [10]:
def train_and_predict_random_forest(X,y): 

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=34, stratify=y)


    model = RandomForestClassifier(n_estimators=100, max_features='sqrt', random_state=34)


    # Treina o modelo
    model.fit(X_train, y_train)
    # Faz previsões no conjunto de teste
    y_pred = model.predict(X_test)
    # Calcula e armazena o relatório de classificação
    report = classification_report(y_test, y_pred, zero_division=0)
    # Calcula e armazena o F1 Macro Score
    f1 = f1_score(y_test, y_pred, average='macro')

    print("Model  (Random_Forest):\n{}".format(report))
    print("F1_Macro: {:.5f}".format(f1))
    print("=" * 50)

    return f1 

def train_and_predict_extra_trees(X,y): 

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=34, stratify=y)


    model =   ExtraTreesClassifier(criterion='gini', max_depth=20, random_state=34)



    # Treina o modelo
    model.fit(X_train, y_train)
    # Faz previsões no conjunto de teste
    y_pred = model.predict(X_test)
    # Calcula e armazena o relatório de classificação
    report = classification_report(y_test, y_pred, zero_division=0)
    # Calcula e armazena o F1 Macro Score
    f1 = f1_score(y_test, y_pred, average='macro')

    print("Model  (Random_Forest):\n{}".format(report))
    print("F1_Macro: {:.5f}".format(f1))
    print("=" * 50)

    return f1 

def train_and_predict_Gradient_Boost(X,y): 

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=34, stratify=y)


    model =     GradientBoostingClassifier(
        learning_rate=0.1, n_estimators=100, random_state=34
    )

    # Treina o modelo
    model.fit(X_train, y_train)
    # Faz previsões no conjunto de teste
    y_pred = model.predict(X_test)
    # Calcula e armazena o relatório de classificação
    report = classification_report(y_test, y_pred, zero_division=0)
    # Calcula e armazena o F1 Macro Score
    f1 = f1_score(y_test, y_pred, average='macro')

    print("Model  (Random_Forest):\n{}".format(report))
    print("F1_Macro: {:.5f}".format(f1))
    print("=" * 50)

    return f1 

def train_and_predict_XGBOOST(X,y): 

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=34, stratify=y)


    model =  XGBClassifier(n_estimators=100, max_depth=4, learning_rate=0.05,colsample_bytree=0.4, subsample=0.8, random_state=34)



    # Treina o modelo
    model.fit(X_train, y_train)
    # Faz previsões no conjunto de teste
    y_pred = model.predict(X_test)
    # Calcula e armazena o relatório de classificação
    report = classification_report(y_test, y_pred, zero_division=0)
    # Calcula e armazena o F1 Macro Score
    f1 = f1_score(y_test, y_pred, average='macro')

    print("Model  (Random_Forest):\n{}".format(report))
    print("F1_Macro: {:.5f}".format(f1))
    print("=" * 50)

    return f1 


### Ler CSV melhor para cada Modelo

In [11]:

data_ExtraTrees = pd.read_csv('2_data_ExtraTrees.csv')
data_Random_Forest = pd.read_csv('2_data_Random_Forest.csv')
data_GradientBoost = pd.read_csv('2_data_GradientBoost.csv')
data_XGBoost = pd.read_csv('2_data_XGBoost.csv')



In [13]:
thresholds = [0.001, 0.002, 0.003, 0.004, 0.005]
all_results = []  # Lista para armazenar os resultados

for threshold in thresholds:
    print(f"Testing threshold: {threshold}\n\n")
    
    result_entry = {
        'threshold': threshold,
        'features_removed': None,  # Atualizado depois
        'remaining_features': None,  # Atualizado depois
        'F1_score_Random_Forest': None,
        'F1_score_Extra_trees': None,
        'F1_score_Gradient_Boost': None,
        'F1_score_XGBOOST': None
    }
    
    # ? ######### RANDOM_FOREST
    X_cleaned, to_drop = remove_low_variance_features(data_Random_Forest, threshold)
    X_Limpo = X_cleaned.drop('Transition', axis=1)
    y = X_cleaned['Transition']
    print(f"Threshold: {threshold}, Features removed: {len(to_drop)}, Remaining features: {X_cleaned.shape[1]}")
    results = train_and_predict_random_forest(X_Limpo, y)
    result_entry['F1_score_Random_Forest'] = results
    result_entry['features_removed'] = len(to_drop)
    result_entry['remaining_features'] = X_cleaned.shape[1]
    print("\n\n\n ##########################################\n\n\n")
    
    # ? ######### EXTRA_TREES
    X_cleaned, to_drop = remove_low_variance_features(data_ExtraTrees, threshold)
    X_Limpo = X_cleaned.drop('Transition', axis=1)
    y = X_cleaned['Transition']
    print(f"Threshold: {threshold}, Features removed: {len(to_drop)}, Remaining features: {X_cleaned.shape[1]}")
    results = train_and_predict_extra_trees(X_Limpo, y)
    result_entry['F1_score_Extra_trees'] = results
    print("\n\n\n ##########################################\n\n\n")
    
    # ? ######### GRADIENT_BOOSTING
    X_cleaned, to_drop = remove_low_variance_features(data_GradientBoost, threshold)
    X_Limpo = X_cleaned.drop('Transition', axis=1)
    y = X_cleaned['Transition']
    print(f"Threshold: {threshold}, Features removed: {len(to_drop)}, Remaining features: {X_cleaned.shape[1]}")
    results = train_and_predict_Gradient_Boost(X_Limpo, y)
    result_entry['F1_score_Gradient_Boost'] = results
    print("\n\n\n ##########################################\n\n\n")
    
    # ? ######### XGBOOST
    X_cleaned, to_drop = remove_low_variance_features(data_XGBoost, threshold)
    X_Limpo = X_cleaned.drop('Transition', axis=1)
    y = X_cleaned['Transition']
    print(f"Threshold: {threshold}, Features removed: {len(to_drop)}, Remaining features: {X_cleaned.shape[1]}")
    results = train_and_predict_XGBOOST(X_Limpo, y)
    result_entry['F1_score_XGBOOST'] = results
    print("\n\n\n ##########################################\n\n\n")
    
    # Adicionar o resultado consolidado
    all_results.append(result_entry)

# Comparar e imprimir os resultados no final
print("\nSummary of Results:")
for res in all_results:
    print(f"Threshold: {res['threshold']}, Features removed: {res['features_removed']}, "
          f"Remaining features: {res['remaining_features']}, "
          f"Random Forest: {res['F1_score_Random_Forest']}, "
          f"Extra Trees: {res['F1_score_Extra_trees']}, "
          f"Gradient Boost: {res['F1_score_Gradient_Boost']}, "
          f"XGBoost: {res['F1_score_XGBOOST']}")


Testing threshold: 0.001


Threshold: 0.001, Features removed: 10, Remaining features: 375
Model  (Random_Forest):
              precision    recall  f1-score   support

           0       0.51      0.83      0.63        24
           1       0.54      0.47      0.50        15
           2       0.33      0.35      0.34        17
           3       0.29      0.11      0.16        18
           4       0.00      0.00      0.00         3

    accuracy                           0.45        77
   macro avg       0.33      0.35      0.33        77
weighted avg       0.41      0.45      0.41        77

F1_Macro: 0.32756



 ##########################################



Threshold: 0.001, Features removed: 33, Remaining features: 779
Model  (Random_Forest):
              precision    recall  f1-score   support

           0       0.56      0.83      0.67        24
           1       0.53      0.53      0.53        15
           2       0.25      0.24      0.24        17
           3       0.30

| Modelo           |   F1_Macro Original                |F1_Macro (0.001) | F1_Macro (0.002) | F1_Macro (0.003) | F1_Macro (0.004) | F1_Macro (0.005) |
|------------------|-------------------|-----------------|------------------|------------------|------------------|------------------|
| Random Forest    |    0.3800               |   0.3276        | 0.3037           | 0.3980           | 0.2766           | **0.4198**           |
| Extra Trees      |     0.3707              |   0.3313        | 0.3168           | 0.2567           | 0.3076           | 0.2759           |
| Gradient Boost   |      0.4097             |   0.4057        | 0.3644           | 0.3421           | 0.3642           | 0.3395           |
| XGBoost          |  0.3705                 |   0.3478        | 0.3452           | 0.3196           | 0.3266           | 0.3214           |

Apenas o Random Forest Demonstrou melhoria, por isso iremos continuar a testar

In [14]:
thresholds = [0.006, 0.007, 0.008, 0.009, 0.01]
all_results = []  # Lista para armazenar os resultados

for threshold in thresholds:
    print(f"Testing threshold: {threshold}\n\n")
    
    result_entry = {
        'threshold': threshold,
        'features_removed': None,  # Atualizado depois
        'remaining_features': None,  # Atualizado depois
        'F1_score_Random_Forest': None,
        'F1_score_Extra_trees': None,
        'F1_score_Gradient_Boost': None,
        'F1_score_XGBOOST': None
    }
    
    # ? ######### RANDOM_FOREST
    X_cleaned, to_drop = remove_low_variance_features(data_Random_Forest, threshold)
    X_Limpo = X_cleaned.drop('Transition', axis=1)
    y = X_cleaned['Transition']
    print(f"Threshold: {threshold}, Features removed: {len(to_drop)}, Remaining features: {X_cleaned.shape[1]}")
    results = train_and_predict_random_forest(X_Limpo, y)
    result_entry['F1_score_Random_Forest'] = results
    result_entry['features_removed'] = len(to_drop)
    result_entry['remaining_features'] = X_cleaned.shape[1]
    print("\n\n\n ##########################################\n\n\n")
    

    # Adicionar o resultado consolidado
    all_results.append(result_entry)

# Comparar e imprimir os resultados no final
print("\nSummary of Results:")
for res in all_results:
    print(f"Threshold: {res['threshold']}, Features removed: {res['features_removed']}, "
          f"Remaining features: {res['remaining_features']}, "
          f"Random Forest: {res['F1_score_Random_Forest']}")


Testing threshold: 0.006


Threshold: 0.006, Features removed: 39, Remaining features: 346
Model  (Random_Forest):
              precision    recall  f1-score   support

           0       0.56      0.79      0.66        24
           1       0.57      0.53      0.55        15
           2       0.30      0.35      0.32        17
           3       0.33      0.17      0.22        18
           4       0.00      0.00      0.00         3

    accuracy                           0.47        77
   macro avg       0.35      0.37      0.35        77
weighted avg       0.43      0.47      0.44        77

F1_Macro: 0.35069



 ##########################################



Testing threshold: 0.007


Threshold: 0.007, Features removed: 44, Remaining features: 341
Model  (Random_Forest):
              precision    recall  f1-score   support

           0       0.50      0.88      0.64        24
           1       0.50      0.40      0.44        15
           2       0.33      0.29      0.31       

Nenhum Threshold melhorou mais a performance do random_forest, portanto para este modelo iremos adicionar o Threshold 0.005 no seu modelo de dados 

In [15]:
X_cleaned, to_drop = remove_low_variance_features(data_Random_Forest, 0.005)

X_cleaned.to_csv('2_data_Random_Forest_Clean.csv', index=False)
